player/js/utils/shapes/ZigZagModifier.js

```diff
@@ -0,0 +1,91 @@
+import {
+  extendPrototype,
+} from '../functionExtensions';
+import PropertyFactory from '../PropertyFactory';
+import shapePool from '../pooling/shape_pool';
+import {
+  ShapeModifier,
+} from './ShapeModifiers';
+import { PolynomialBezier } from '../PolynomialBezier';
+
+function ZigZagModifier() {}
+extendPrototype([ShapeModifier], ZigZagModifier);
+ZigZagModifier.prototype.initModifierProperties = function (elem, data) {
+  this.getValue = this.processKeys;
+  this.amplitude = PropertyFactory.getProp(elem, data.s, 0, null, this);
+  this.frequency = PropertyFactory.getProp(elem, data.pt, 0, null, this);
+  this._isAnimated = this.amplitude.effectsSequence.length !== 0 && this.frequency.effectsSequence.length !== 0;
+};
+
+ZigZagModifier.prototype.processPath = function (path, amplitude, frequency) {
+  var pathLength = path._length;
+  var clonedPath = shapePool.newElement();
+  clonedPath.c = path.c;
+  var vX;
+  var vY;
+  var direction = 1;
+
+  if (!path.c) {
+    pathLength -= 1;
+  }
+
+  for (var cur = 0; cur < pathLength; cur += 1) {
+    var bez = PolynomialBezier.shapeSegment(path, cur);
+
+    vX = bez.points[0][0];
+    vY = bez.points[0][1];
+    clonedPath.setTripleAt(vX, vY, vX, vY, vX, vY, clonedPath.length());
```

Comment:
Comparing the behavior of AE with the result of this calculation, it seems that this point should also be offset by the amplitude.  
I'm looking into how to the tangent is calculated for it.

Reply:
After some more tests, I figured out that the vertices need to be offset by a vector tangent to its surrounding vertices.

Reply:
Offset by the bezier tangent?